File: app/src/main/java/com/baidu/tv/player/auth/BaiduAuthService.java
package com.baidu.tv.player.auth;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Handler;
import android.os.Looper;

import com.baidu.tv.player.config.BaiduConfig;
import com.baidu.tv.player.model.AuthInfo;
import com.baidu.tv.player.model.DeviceCodeResponse;
import com.baidu.tv.player.model.TokenResponse;
import com.baidu.tv.player.model.UserInfoResponse;
import com.baidu.tv.player.network.ApiConstants;
import com.baidu.tv.player.network.BaiduPanService;
import com.baidu.tv.player.network.RetrofitClient;

import java.io.IOException;
import java.util.UUID;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * 百度网盘认证服务
 */
public class BaiduAuthService {
    
    private static final String PREF_NAME = "baidu_auth";
    private static final String KEY_ACCESS_TOKEN = "access_token";
    private static final String KEY_REFRESH_TOKEN = "refresh_token";
    private static final String KEY_EXPIRES_AT = "expires_at";
    private static final String KEY_SCOPE = "scope";
    private static final String KEY_SESSION_KEY = "session_key";
    private static final String KEY_SESSION_SECRET = "session_secret";
    private static final String KEY_SESSION_EXPIRES_AT = "session_expires_at";
    private static final String KEY_USER_ID = "user_id";
    private static final String KEY_USERNAME = "username";
    private static final String KEY_IS_LOGGED_IN = "is_logged_in";
    private static final String KEY_DEVICE_ID = "device_id";
    
    private static volatile BaiduAuthService instance;
    private final Context context;
    private final SharedPreferences prefs;
    private final BaiduPanService oauthService;
    private AuthInfo authInfo;
    private Handler handler;
    
    private BaiduAuthService(Context context) {
        this.context = context.getApplicationContext();
        this.prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        this.oauthService = RetrofitClient.getOAuthInstance().create(BaiduPanService.class);
        this.handler = new Handler(Looper.getMainLooper());
        this.authInfo = loadAuthInfo();
    }
    
    public static BaiduAuthService getInstance(Context context) {
        if (instance == null) {
            synchronized (BaiduAuthService.class) {
                if (instance == null) {
                    instance = new BaiduAuthService(context);
                }
            }
        }
        return instance;
    }
    
    /**
     * 从SharedPreferences加载认证信息
     */
    private AuthInfo loadAuthInfo() {
        AuthInfo info = new AuthInfo();
        info.setAccessToken(prefs.getString(KEY_ACCESS_TOKEN, ""));
        info.setRefreshToken(prefs.getString(KEY_REFRESH_TOKEN, ""));
        info.setExpiresAt(prefs.getLong(KEY_EXPIRES_AT, 0));
        info.setScope(prefs.getString(KEY_SCOPE, ""));
        info.setSessionKey(prefs.getString(KEY_SESSION_KEY, ""));
        info.setSessionSecret(prefs.getString(KEY_SESSION_SECRET, ""));
        info.setSessionExpiresAt(prefs.getLong(KEY_SESSION_EXPIRES_AT, 0));
        info.setUserId(prefs.getString(KEY_USER_ID, ""));
        info.setUsername(prefs.getString(KEY_USERNAME, ""));
        info.setLoggedIn(prefs.getBoolean(KEY_IS_LOGGED_IN, false));
        return info;
    }
    
    /**
     * 保存认证信息到SharedPreferences
     */
    private void saveAuthInfo() {
        SharedPreferences.Editor editor = prefs.edit();
        editor.putString(KEY_ACCESS_TOKEN, authInfo.getAccessToken());
        editor.putString(KEY_REFRESH_TOKEN, authInfo.getRefreshToken());
        editor.putLong(KEY_EXPIRES_AT, authInfo.getExpiresAt());
        editor.putString(KEY_SCOPE, authInfo.getScope());
        editor.putString(KEY_SESSION_KEY, authInfo.getSessionKey());
        editor.putString(KEY_SESSION_SECRET, authInfo.getSessionSecret());
        editor.putLong(KEY_SESSION_EXPIRES_AT, authInfo.getSessionExpiresAt());
        editor.putString(KEY_USER_ID, authInfo.getUserId());
        editor.putString(KEY_USERNAME, authInfo.getUsername());
        editor.putBoolean(KEY_IS_LOGGED_IN, authInfo.isLoggedIn());
        editor.apply();
    }
    
    /**
     * 获取或生成设备ID
     */
    public String getDeviceId() {
        String deviceId = prefs.getString(KEY_DEVICE_ID, null);
        if (deviceId == null) {
            deviceId = UUID.randomUUID().toString();
            prefs.edit().putString(KEY_DEVICE_ID, deviceId).apply();
        }
        return deviceId;
    }
    
    /**
     * 检查是否已认证
     */
    public boolean isAuthenticated() {
        if (!authInfo.isLoggedIn()) {
            return false;
        }
        
        if (authInfo.getAccessToken() == null || authInfo.getAccessToken().isEmpty()) {
            return false;
        }
        
        // 检查token是否过期
        if (System.currentTimeMillis() >= authInfo.getExpiresAt()) {
            return false;
        }
        
        return true;
    }
    
    /**
     * 获取访问令牌
     */
    public String getAccessToken() {
        return authInfo.getAccessToken();
    }
    
    /**
     * 获取认证信息
     */
    public AuthInfo getAuthInfo() {
        return authInfo;
    }
    
    /**
     * 获取设备码
     */
    public void getDeviceCode(AuthCallback<DeviceCodeResponse> callback) {
        Call<DeviceCodeResponse> call = oauthService.getDeviceCode(
                BaiduConfig.APP_KEY,
                BaiduConfig.SCOPE,
                "device_code"
        );
        
        call.enqueue(new Callback<DeviceCodeResponse>() {
            @Override
            public void onResponse(Call<DeviceCodeResponse> call, Response<DeviceCodeResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    DeviceCodeResponse deviceCodeResponse = response.body();
                    if (deviceCodeResponse.hasError()) {
                        callback.onError(deviceCodeResponse.getError());
                    } else {
                        callback.onSuccess(deviceCodeResponse);
                    }
                } else {
                    callback.onError("获取设备码失败");
                }
            }
            
            @Override
            public void onFailure(Call<DeviceCodeResponse> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }
    
    /**
     * 轮询设备码状态
     */
    public void pollDeviceCodeStatus(String deviceCode, AuthCallback<Boolean> callback) {
        pollDeviceCodeStatus(deviceCode, 0, callback);
    }
    
    private void pollDeviceCodeStatus(String deviceCode, int count, AuthCallback<Boolean> callback) {
        if (count >= ApiConstants.MAX_POLLING_COUNT) {
            callback.onError("授权超时");
            return;
        }
        
        Call<TokenResponse> call = oauthService.getTokenByDeviceCode(
                "device_token",
                deviceCode,
                BaiduConfig.APP_KEY,
                BaiduConfig.SECRET_KEY
        );
        
        call.enqueue(new Callback<TokenResponse>() {
            @Override
            public void onResponse(Call<TokenResponse> call, Response<TokenResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    TokenResponse tokenResponse = response.body();
                    
                    // 授权成功
                    if (tokenResponse.getAccessToken() != null && !tokenResponse.getAccessToken().isEmpty()) {
                        authInfo.setAccessToken(tokenResponse.getAccessToken());
                        authInfo.setRefreshToken(tokenResponse.getRefreshToken());
                        authInfo.setExpiresAt(System.currentTimeMillis() + tokenResponse.getExpiresIn() * 1000);
                        authInfo.setScope(tokenResponse.getScope());
                        authInfo.setSessionKey(tokenResponse.getSessionKey());
                        authInfo.setSessionSecret(tokenResponse.getSessionSecret());
                        authInfo.setLoggedIn(true);
                        saveAuthInfo();
                        callback.onSuccess(true);
                    } else {
                        // 理论上成功响应不应该包含错误，但为了健壮性保留处理
                        handleAuthError(tokenResponse.getError(), tokenResponse.getErrorDescription(), deviceCode, count, callback);
                    }
                } else {
                    // 处理4xx错误，Baidu API在pending状态会返回400
                    try {
                        String errorBody = response.errorBody().string();
                        // 简单的JSON解析，避免引入额外依赖
                        if (errorBody.contains("authorization_pending")) {
                            handler.postDelayed(() -> {
                                pollDeviceCodeStatus(deviceCode, count + 1, callback);
                            }, ApiConstants.POLLING_INTERVAL);
                        } else if (errorBody.contains("expired_token")) {
                            callback.onError("授权已过期");
                        } else {
                            callback.onError("授权失败: " + response.code());
                        }
                    } catch (Exception e) {
                        callback.onError("授权失败: " + e.getMessage());
                    }
                }
            }

            private void handleAuthError(String error, String description, String deviceCode, int count, AuthCallback<Boolean> callback) {
                // 授权中
                if ("authorization_pending".equals(error)) {
                    handler.postDelayed(() -> {
                        pollDeviceCodeStatus(deviceCode, count + 1, callback);
                    }, ApiConstants.POLLING_INTERVAL);
                }
                // 授权过期
                else if ("expired_token".equals(error)) {
                    callback.onError("授权已过期");
                }
                // 其他错误
                else {
                    callback.onError(description != null ? description : "授权失败");
                }
            }
            
            @Override
            public void onFailure(Call<TokenResponse> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }
    
    /**
     * 刷新token
     */
    public void refreshToken(AuthCallback<Boolean> callback) {
        if (authInfo.getRefreshToken() == null || authInfo.getRefreshToken().isEmpty()) {
            callback.onError("无刷新令牌");
            return;
        }
        
        Call<TokenResponse> call = oauthService.refreshToken(
                "refresh_token",
                authInfo.getRefreshToken(),
                BaiduConfig.APP_KEY,
                BaiduConfig.SECRET_KEY
        );
        
        call.enqueue(new Callback<TokenResponse>() {
            @Override
            public void onResponse(Call<TokenResponse> call, Response<TokenResponse> response) {
                if (response.isSuccessful() && response.body() != null) {
                    TokenResponse tokenResponse = response.body();
                    
                    if (tokenResponse.getAccessToken() != null && !tokenResponse.getAccessToken().isEmpty()) {
                        authInfo.setAccessToken(tokenResponse.getAccessToken());
                        authInfo.setRefreshToken(tokenResponse.getRefreshToken());
                        authInfo.setExpiresAt(System.currentTimeMillis() + tokenResponse.getExpiresIn() * 1000);
                        authInfo.setScope(tokenResponse.getScope());
                        authInfo.setLoggedIn(true);
                        saveAuthInfo();
                        callback.onSuccess(true);
                    } else {
                        callback.onError(tokenResponse.getErrorDescription() != null ? 
                                tokenResponse.getErrorDescription() : "刷新令牌失败");
                    }
                } else {
                    callback.onError("刷新令牌失败");
                }
            }
            
            @Override
            public void onFailure(Call<TokenResponse> call, Throwable t) {
                callback.onError(t.getMessage());
            }
        });
    }
    
    /**
     * 退出登录
     */
    public void logout() {
        authInfo = new AuthInfo();
        saveAuthInfo();
    }
    
    /**
     * 认证回调接口
     */
    public interface AuthCallback<T> {
        void onSuccess(T result);
        void onError(String error);
    }
}

File: app/src/main/java/com/baidu/tv/player/kt/auth/BaiduAuthService.kt
package com.baidu.tv.player.kt.auth

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import com.baidu.tv.player.kt.config.BaiduConfig
import com.baidu.tv.player.kt.di.AuthPrefs
import com.baidu.tv.player.kt.di.OAuthApi
import com.baidu.tv.player.kt.model.AuthInfo
import com.baidu.tv.player.kt.model.DeviceCodeResponse
import com.baidu.tv.player.kt.model.TokenResponse
import com.baidu.tv.player.kt.network.ApiConstants
import com.baidu.tv.player.kt.network.BaiduPanService
import com.google.gson.Gson
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.delay
import retrofit2.HttpException
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "BaiduAuthService"

/**
 * 百度网盘认证服务。
 *
 * - 由 Hilt 注入，替代 Java 版手动单例。
 * - Token 读写使用 [EncryptedSharedPreferences]（AES-256 GCM）。
 * - 全部为 `suspend` 函数，协程 `delay` 替代 Handler.postDelayed 轮询。
 *
 * 设备码授权轮询：百度在 pending 状态返回 HTTP 400 + `{"error":"authorization_pending"}`，
 * 成功返回 200 + token JSON。suspend Retrofit 在 4xx 抛 [HttpException]，需解析 errorBody。
 */
@Singleton
class BaiduAuthService @Inject constructor(
    @ApplicationContext private val context: Context,
    @OAuthApi private val oauthService: BaiduPanService,
    @AuthPrefs private val prefs: SharedPreferences,
) {
    private var authInfo: AuthInfo = loadAuthInfo()

    /** OAuth 授权状态（轮询用）。 */
    enum class PollResult { AUTHENTICATED, PENDING, EXPIRED, ERROR }

    /** 检查是否已认证。 */
    fun isAuthenticated(): Boolean {
        if (!authInfo.isLoggedIn) return false
        if (authInfo.accessToken.isNullOrEmpty()) return false
        if (System.currentTimeMillis() >= authInfo.expiresAt) return false
        return true
    }

    /** 获取访问令牌（仅返回当前仍在有效期内的 token）。 */
    fun getAccessToken(): String? = authInfo.accessToken?.takeIf { isAuthenticated() }

    /** 获取认证信息。 */
    fun getAuthInfo(): AuthInfo = authInfo

    /** 获取或生成设备 ID（持久化）。 */
    fun getDeviceId(): String {
        prefs.getString(KEY_DEVICE_ID, null)?.let { return it }
        val id = UUID.randomUUID().toString()
        prefs.edit().putString(KEY_DEVICE_ID, id).apply()
        return id
    }

    /** 获取设备码。失败时抛异常。 */
    suspend fun getDeviceCode(): DeviceCodeResponse {
        val resp = oauthService.getDeviceCode(BaiduConfig.APP_KEY, BaiduConfig.SCOPE, "device_code")
        if (resp.hasError()) throw IllegalStateException(resp.error)
        return resp
    }

    /**
     * 轮询设备码状态。成功保存 token 并返回 [PollResult.AUTHENTICATED]；
     * pending 返回 [PollResult.PENDING]（调用方按 [ApiConstants.POLLING_INTERVAL] 间隔重试）；
     * 过期返回 [PollResult.EXPIRED]；其他错误返回 [PollResult.ERROR]。
     */
    suspend fun pollDeviceCodeStatus(deviceCode: String): PollResult {
        return try {
            val token = oauthService.getTokenByDeviceCode(
                "device_token", deviceCode, BaiduConfig.APP_KEY, BaiduConfig.SECRET_KEY,
            )
            if (!token.accessToken.isNullOrEmpty()) {
                applyToken(token)
                PollResult.AUTHENTICATED
            } else if (token.error == "authorization_pending") {
                PollResult.PENDING
            } else if (token.error == "expired_token") {
                PollResult.EXPIRED
            } else {
                Log.e(TAG, "轮询返回错误: ${token.error} - ${token.errorDescription}")
                PollResult.ERROR
            }
        } catch (e: HttpException) {
            // 4xx：解析 errorBody 中的 error 字段
            val errorBody = try { e.response()?.errorBody()?.string() } catch (_: Exception) { null }
            val error = parseError(errorBody)
            when (error) {
                "authorization_pending" -> PollResult.PENDING
                "expired_token" -> PollResult.EXPIRED
                "slow_down" -> { delay(ApiConstants.POLLING_INTERVAL); PollResult.PENDING }
                else -> {
                    Log.e(TAG, "轮询 HTTP 错误: ${e.code()}, body=$errorBody")
                    PollResult.ERROR
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "轮询网络错误", e)
            PollResult.ERROR
        }
    }

    /**
     * 完整的设备码轮询循环：直到认证成功 / 过期 / 出错 / 达到最大次数。
     * @return 终态 [PollResult]：[PollResult.AUTHENTICATED] 成功；
     * [PollResult.EXPIRED] 设备码过期（含轮询次数耗尽，调用方可刷新二维码重试）；
     * [PollResult.ERROR] 不可恢复错误。
     */
    suspend fun pollUntilAuthenticated(deviceCode: String): PollResult {
        repeat(ApiConstants.MAX_POLLING_COUNT) {
            when (val result = pollDeviceCodeStatus(deviceCode)) {
                PollResult.PENDING -> delay(ApiConstants.POLLING_INTERVAL)
                else -> return result
            }
        }
        // 轮询次数耗尽视为设备码过期，由调用方刷新二维码
        return PollResult.EXPIRED
    }

    /** 刷新 token。成功返回 true。 */
    suspend fun refreshToken(): Boolean {
        if (authInfo.refreshToken.isNullOrEmpty()) return false
        return try {
            val token = oauthService.refreshToken(
                "refresh_token", authInfo.refreshToken!!, BaiduConfig.APP_KEY, BaiduConfig.SECRET_KEY,
            )
            if (!token.accessToken.isNullOrEmpty()) {
                applyToken(token)
                true
            } else {
                Log.e(TAG, "刷新令牌失败: ${token.errorDescription}")
                false
            }
        } catch (e: Exception) {
            Log.e(TAG, "刷新令牌异常", e)
            false
        }
    }

    /** 撤销 token（登出，服务端）。 */
    suspend fun revokeToken() {
        try {
            authInfo.accessToken?.let { oauthService.revokeToken(it) }
        } catch (e: Exception) {
            Log.w(TAG, "撤销 token 失败（忽略）", e)
        }
    }

    /** 退出登录：清空内存与持久化认证信息。 */
    fun logout() {
        authInfo = AuthInfo()
        saveAuthInfo()
    }

    /**
     * 迁移专用入口：用已读取的旧明文 [AuthInfo] 覆盖内存态并持久化到加密存储。
     * 仅供 [AuthMigration] 一次性迁移使用。
     */
    fun applyMigratedAuthInfo(info: AuthInfo) {
        authInfo = info
        saveAuthInfo()
    }

    /** 将 [TokenResponse] 写入内存 authInfo 并持久化。 */
    private fun applyToken(token: TokenResponse) {
        authInfo.accessToken = token.accessToken
        authInfo.refreshToken = token.refreshToken
        authInfo.expiresAt = System.currentTimeMillis() + token.expiresIn * 1000L
        authInfo.scope = token.scope
        authInfo.sessionKey = token.sessionKey
        authInfo.sessionSecret = token.sessionSecret
        authInfo.isLoggedIn = true
        saveAuthInfo()
    }

    private fun loadAuthInfo(): AuthInfo = AuthInfo().apply {
        accessToken = prefs.getString(KEY_ACCESS_TOKEN, "")
        refreshToken = prefs.getString(KEY_REFRESH_TOKEN, "")
        expiresAt = prefs.getLong(KEY_EXPIRES_AT, 0)
        scope = prefs.getString(KEY_SCOPE, "")
        sessionKey = prefs.getString(KEY_SESSION_KEY, "")
        sessionSecret = prefs.getString(KEY_SESSION_SECRET, "")
        sessionExpiresAt = prefs.getLong(KEY_SESSION_EXPIRES_AT, 0)
        userId = prefs.getString(KEY_USER_ID, "")
        username = prefs.getString(KEY_USERNAME, "")
        isLoggedIn = prefs.getBoolean(KEY_IS_LOGGED_IN, false)
    }

    private fun saveAuthInfo() {
        prefs.edit().apply {
            putString(KEY_ACCESS_TOKEN, authInfo.accessToken)
            putString(KEY_REFRESH_TOKEN, authInfo.refreshToken)
            putLong(KEY_EXPIRES_AT, authInfo.expiresAt)
            putString(KEY_SCOPE, authInfo.scope)
            putString(KEY_SESSION_KEY, authInfo.sessionKey)
            putString(KEY_SESSION_SECRET, authInfo.sessionSecret)
            putLong(KEY_SESSION_EXPIRES_AT, authInfo.sessionExpiresAt)
            putString(KEY_USER_ID, authInfo.userId)
            putString(KEY_USERNAME, authInfo.username)
            putBoolean(KEY_IS_LOGGED_IN, authInfo.isLoggedIn)
        }.apply()
    }

    /** 从 errorBody JSON 解析 `error` 字段。 */
    private fun parseError(errorBody: String?): String? = try {
        if (errorBody.isNullOrEmpty()) null
        else Gson().fromJson(errorBody, TokenResponse::class.java).error
    } catch (e: Exception) {
        null
    }

    companion object {
        private const val KEY_ACCESS_TOKEN = "access_token"
        private const val KEY_REFRESH_TOKEN = "refresh_token"
        private const val KEY_EXPIRES_AT = "expires_at"
        private const val KEY_SCOPE = "scope"
        private const val KEY_SESSION_KEY = "session_key"
        private const val KEY_SESSION_SECRET = "session_secret"
        private const val KEY_SESSION_EXPIRES_AT = "session_expires_at"
        private const val KEY_USER_ID = "user_id"
        private const val KEY_USERNAME = "username"
        private const val KEY_IS_LOGGED_IN = "is_logged_in"
        private const val KEY_DEVICE_ID = "device_id"

        /** 旧明文 SharedPreferences 文件名（用于一次性迁移检测）。 */
        const val LEGACY_PLAIN_PREF_NAME = "baidu_aid"
    }
}
